File: Part 4 - Clustering/Section 24 - K-Means Clustering/my_model.R
dataset <- read.csv('Mall_Customers.csv')
X <- dataset[4:5]

# elbow method to find number of clusters
set.seed(6)
wcss <- vector()
for (i in 1:10) wcss[i] <- sum(kmeans(X, i)$withinss)
plot(1:10, wcss, type = "b", main = paste('clusters of clients'), xlab = 'number of clusters', ylab = 'wcss')

# applying k-means to mall data
set.seed(29)
kmeans <- kmeans(X, 5, iter.max = 300, nstart = 10)

# vis clusters
library(cluster)
clusplot(X, 
         kmeans$cluster, 
         lines = 0, 
         shade = TRUE, 
         color = TRUE, 
         labels = 2, 
         plotchar = FALSE,
         span = TRUE,
         main = paste('clusters of clients'),
         xlab = 'annual income',
         ylab = 'spending score')

File: Part 4 - Clustering/Section 24 - K-Means Clustering/my_model.py
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt

dataset = pd.read_csv('Mall_Customers.csv')
X = dataset.iloc[:, [3, 4]].values

# elbow method to find number of clusters
from sklearn.cluster import KMeans
# wcss = []
# for i in range(1, 11):
#     kmeans = KMeans(n_clusters=i, init='k-means++', max_iter=300, n_init=10, random_state=0)
#     kmeans.fit(X)
#     wcss.append(kmeans.inertia_)
# plt.plot(range(1, 11), wcss)
# plt.title('Elbow method')
# plt.xlabel('#clusters')
# plt.ylabel('wcss')
# plt.show()

# apply k-means to mall dataset
kmeans = KMeans(n_clusters=5, init='k-means++', max_iter=300, n_init=10, random_state=0)
y_kmeans = kmeans.fit_predict(X)

# vis clusters
plt.scatter(X[:, 0], X[:, 1], c=y_kmeans)
plt.show()
